caffe2/core/context_gpu.h

```diff
@@ -334,26 +334,26 @@ inline void CPUContext::CopyBytes<CPUContext, CUDAContext>(
  * GPU present during runtime, at global initialization time we will set
  * the CPU memory allocator to allocate pinned memory.
  */
-struct CAFFE2_CUDA_API PinnedCPUAllocator final : CPUAllocator {
+struct CAFFE2_CUDA_API PinnedCPUAllocator final : public at::Allocator {
   PinnedCPUAllocator() {}
   ~PinnedCPUAllocator() override {}
-  std::pair<void*, MemoryDeleter> New(size_t nbytes) override {
+  at::DataPtr allocate(size_t nbytes) const override {
     void* data;
     std::lock_guard<std::mutex> lock(CUDAContext::mutex());
     if (IsNUMAEnabled()) {
-      auto ptr_and_deleter = baseAllocator_.New(nbytes);
-      data = ptr_and_deleter.first;
+      auto data_ptr = baseAllocator_.allocate(nbytes);
+      data = data_ptr.move_context().release();
```

Comment:
This looks bad; you need to hold on to the old deleter and then repackage the final pointer with it. Honestly, I would just not `release()` at all; just move the context out, and then move it back in for the reconstruction. (If the problem is that we don't have a `DataPtr` constructor that takes a `unique_ptr` context, we should add one.)

Reply:
In this case I'll just do `data = data_ptr.get()` to get the data pointer. let's see if we need a constructor for other places..

Reply:
Err, just getting the data pointer won't preserve ownership correctly, will it?